services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/AsyncBufferingSubscriber.java

```diff
@@ -65,89 +57,40 @@ public void onSubscribe(Subscription subscription) {
 
     @Override
     public void onNext(T item) {
-        storingSubscriber.onNext(item);
-        flushBufferIfNeeded();
-    }
-
-    private void flushBufferIfNeeded() {
-        if (isDelivering.compareAndSet(false, true)) {
-            try {
-                Optional<StoringSubscriber.Event<T>> next = storingSubscriber.peek();
-                while (numRequestsInFlight.get() < maxConcurrentExecutions) {
-                    if (!next.isPresent()) {
-                        subscription.request(1);
-                        break;
-                    }
-
-                    switch (next.get().type()) {
-                        case ON_COMPLETE:
-                            handleCompleteEvent();
-                            break;
-                        case ON_ERROR:
-                            handleError(next.get().runtimeError());
-                            break;
-                        case ON_NEXT:
-                            handleOnNext(next.get().value());
-                            break;
-                        default:
-                            handleError(new IllegalStateException("Unknown stored type: " + next.get().type()));
-                            break;
-                    }
-
-                    next = storingSubscriber.peek();
-                }
-            } finally {
-                isDelivering.set(false);
-            }
-        }
-    }
-
-    private void handleOnNext(T item) {
-        storingSubscriber.poll();
-
-        int numberOfRequestInFlight = numRequestsInFlight.incrementAndGet();
-        log.debug(() -> "Delivering next item, numRequestInFlight=" + numberOfRequestInFlight);
-
+        numRequestsInFlight.incrementAndGet();
         consumer.apply(item).whenComplete((r, t) -> {
             numRequestsInFlight.decrementAndGet();
-            if (!isStreamingDone) {
-                subscription.request(1);
-            } else {
-                flushBufferIfNeeded();
-            }
+            checkForCompletion();
         });
     }
 
-    private void handleCompleteEvent() {
-        if (numRequestsInFlight.get() == 0) {
-            returnFuture.complete(null);
-            storingSubscriber.poll();
-        }
-    }
-
     @Override
     public void onError(Throwable t) {
-        handleError(t);
-        storingSubscriber.onError(t);
-    }
-
-    private void handleError(Throwable t) {
+        upstreamDone = true;
         returnFuture.completeExceptionally(t);
-        storingSubscriber.poll();
     }
 
     @Override
     public void onComplete() {
-        isStreamingDone = true;
-        storingSubscriber.onComplete();
-        flushBufferIfNeeded();
+        upstreamDone = true;
+        checkForCompletion();
+    }
+
+    private void checkForCompletion() {
+        if (!upstreamDone) {
+            subscription.request(1);
+            return;
+        }
+
+        if (upstreamDone && numRequestsInFlight.get() == 0) {
+            returnFuture.complete(null);
```

Comment:
Worth adding a comment to this if statement body: This could get invoked multiple times, but it doesn't matter because future.complete is idempotent. 

Worth adding the comment so that other people don't add code here assuming it only happens once.